### ISSUE TITLE
Custom LSTMCell has cell state gradients being zeros

### ISSUE THREAD

**Issue by DagonArises** [tensorflow/tensorflow#54817]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):  Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Big Sur 11.3.1
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.6.0
- Python version: 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Let c_t be the cell state at step t, c_0 the initial cell state of the same shape. I am trying to compute dc_t/dc_0 for every step as an indicative statistic for the extent of vanishing gradients in LSTM. On a text comment dataset, where each text sequence is labeled with a binary index as target, I have constructed an LSTM and computed and plotted the gradients over steps. 

For the first 20 steps the gradients are decreasing drastically, demonstrating behaviour of vanishing gradients. For the last 20 steps they are completely zero. 

In the reproducible example below I have used random uniform inputs. The same LSTM model and gradient computation process return me the exact SAME behaviour as it behaves on the text data. There is probably something wrong with my code that the LSTM is missing something when computing the gradients.

**Describe the expected behavior**
The above-described gradients dc_t/dc_0 should stay away from zero, as theory of LSTM suggests that it is resistant against vanishing gradients. i.e. LSTM can capture (moderately) long-term dependencies, so dc_t/dc_0 with t being a few hundred steps should not vanish to zero so drastically. Of course longer temporal-dependency e.g. t = 1000 will cause this gradient to vanish. Models like transformers will do better.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
import numpy as np
from numpy import array, asarray, zeros
import pandas as pd 
from tqdm import tqdm
from sklearn.model_selection import train_test_split

import tensorflow as tf
from keras import Input, Model
from keras.models import Sequential
from keras.layers.recurrent import LSTM, GRU, SimpleRNN
from keras.layers.core import Dense, Activation, Dropout, Flatten
from keras.layers.embeddings import Embedding
from sklearn import preprocessing, decomposition, model_selection, metrics, pipeline
from keras.layers import RNN, LSTMCell, Flatten, Bidirectional, SpatialDropout1D
from keras.preprocessing import sequence, text
from keras.callbacks import EarlyStopping, LambdaCallback, TensorBoard
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from keras import backend as k
import os 


# For each c_t, compute gradient dc_t/dc_0:
batch_size = 100; input_length_m = 20
xtr_pad = tf.random.uniform((batch_size*2, input_length_m), maxval = 500, dtype=tf.int32)
ytr = tf.random.normal((batch_size*2, input_length_m, 200))

inp= Input(batch_shape= (batch_size, input_length_m), name= 'input') 
emb_out= Embedding(500, 100, input_length= input_length_m, trainable= False, name= 'embedding')(inp)

class LSTMCellwithStates(LSTMCell):
    def call(self, inputs, states, training=None):
        real_inputs = inputs[:,:self.units] # decouple [h, c]
        outputs, [h,c] = super().call(real_inputs, states, training=training)
        return tf.concat([h, c], axis=1), [h,c]
    
rnn = RNN(LSTMCellwithStates(200), return_sequences= True, return_state= False, name= 'LSTM') 

h0 = tf.Variable(tf.random.uniform((batch_size, 200)))
c0 = tf.Variable(tf.random.uniform((batch_size, 200)))
rnn_allstates= rnn(emb_out, initial_state=[h0, c0])  

model_lstm_mod = Model(inputs=inp, outputs= rnn_allstates[:, :, 200:], name= 'model_LSTMCell')
model_lstm_mod.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])

### Compute gradients: 

ds = tf.data.Dataset.from_tensor_slices((xtr_pad, ytr)).batch(100)

@tf.function
# Compute gradients
def compute_dct_dc0(t, x, c0):
    return tf.gradients(model_lstm_mod(x)[:,t,:], c0)

n_b = int(xtr_pad.shape[0]/ 100)  # 200 batches
n_steps = 20   # look up only the first and last 20 steps

dctdc0_all= tf.zeros([n_b, n_steps])
for b, (x_batch_train, y_batch_train) in enumerate(ds):  # batches 0,1
    grad_batch= []   # a list of 1403 scalar gradients on the current batch
    for t in range(n_steps):  
        # steps 0,...,19
        dctdc0_b_t = compute_dct_dc0(t, x_batch_train, c0)  # (batch_size, n_units)
        grad_t = tf.reduce_mean(abs(dctdc0_b_t[0]), [0,1]) # Scalar dctdc0 at the current batch and step
        print('step', t+1, 'of batch' ,b+1, 'done')
        grad_batch.append(grad_t)
    
    dctdc0_all= tf.concat([dctdc0_all, [grad_batch]], axis = 0)   

dctdc0_agg= tf.reduce_mean(dctdc0_all, 0)  # take rowmean to obtain a vector of shape (20,)
print(dctdc0_agg.shape)

```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The LSTM gradient plot for the reproducible eg:

<img width="647" alt="Screen Shot 2022-03-02 at 4 31 31 pm" src="https://user-images.githubusercontent.com/96099256/156301257-7df3a396-fe2a-4122-be3a-51f90a4a1774.png">

Plots for first and last 20 steps on text data:
<img width="620" alt="Screen Shot 2022-03-02 at 4 29 00 pm" src="https://user-images.githubusercontent.com/96099256/156301342-370fb553-40c8-425a-9963-fd5da2d53199.png">

<img width="617" alt="Screen Shot 2022-03-02 at 4 33 00 pm" src="https://user-images.githubusercontent.com/96099256/156301388-7119d512-3333-4a84-b0a5-70a3710fbd97.png">

**Comment by DagonArises** [comment]
@mohantym done it's [here](https://github.com/keras-team/keras/issues/16151)

**Comment by mohantym** [comment]
Ok! @DagonArises ! Moving this issue to closed status as it will be tracked down in Keras repo. Thanks!
